visualization/app/codeCharta/ui/ribbonBar/ribbonBar.component.ts

```diff
@@ -1,18 +1,41 @@
 import "./ribbonBar.component.scss"
+import { IRootScopeService } from "angular"
 import $ from "jquery"
+import { StoreService } from "../../state/store.service"
+import { PanelSelection } from "../../codeCharta.model"
+import { setPanelSelection } from "../../state/store/appSettings/panelSelection/panelSelection.actions"
+import { PanelSelectionService, PanelSelectionSubscriber } from "../../state/store/appSettings/panelSelection/panelSelection.service"
+import { CodeChartaMouseEventService } from "../../codeCharta.mouseEvent.service"
 
-export class RibbonBarController {
-	private readonly EXPANDED_CLASS = "expanded"
+export class RibbonBarController implements PanelSelectionSubscriber {
+	constructor(
+		private storeService: StoreService,
+		private $rootScope: IRootScopeService,
+		private codeChartaMouseEventService: CodeChartaMouseEventService
+	) {
+		PanelSelectionService.subscribe(this.$rootScope, this)
+	}
 
-	public toggle(event) {
-		const boxElement = $(event.srcElement).closest("md-card")
-		$(document.activeElement).blur()
+	private _viewModel: {
+		panelSelection: PanelSelection
+		panelSelectionValues: typeof PanelSelection
+	} = {
+		panelSelection: PanelSelection.NONE,
+		panelSelectionValues: PanelSelection
+	}
 
-		if (boxElement.hasClass(this.EXPANDED_CLASS)) {
-			boxElement.removeClass(this.EXPANDED_CLASS)
+	public onPanelSelectionChanged(panelSelection: PanelSelection) {
+		this._viewModel.panelSelection = panelSelection
+	}
+
+	public toggle(panelSelection: PanelSelection) {
+		$(document.activeElement).blur()
+		if (this._viewModel.panelSelection !== panelSelection) {
+			this.storeService.dispatch(setPanelSelection(panelSelection))
 		} else {
-			boxElement.addClass(this.EXPANDED_CLASS)
+			this.storeService.dispatch(setPanelSelection(PanelSelection.NONE))
 		}
```

Comment:
This could also be written as:

```js
const selected = this.#viewModel.panelSelection !== panelSelection ? PanelSelection.NONE : panelSelection
this.storeService.dispatch(setPanelSelection(selected))
```

If you want to prevent unnecessary dispatch operations you could also check if both are already `NONE`. It's probably fine as it is though.

Reply:
Your suggested ternary expression is wrong. It should be

```ts
const newSelection = this._viewModel.panelSelection !== panelSelection ? panelSelection : PanelSelection.NONE
this.storeService.dispatch(setPanelSelection(newSelection))
```

There are no unnecessary dispatches. The toggle is triggered by the user and a toggle always triggers an event, either close or open.